spring-context-support/src/main/java/org/springframework/cache/transaction/TransactionAwareCache.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2002-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cache.transaction;
+
+import java.util.concurrent.Callable;
+
+import org.springframework.cache.Cache;
+import org.springframework.lang.Nullable;
+import org.springframework.transaction.support.TransactionSynchronizationAdapter;
+import org.springframework.transaction.support.TransactionSynchronizationManager;
+import org.springframework.util.Assert;
+
+/**
+ * Cache decorator which synchronizes its {@link #put}, {@link #evict} and {@link #clear}
+ * operations with Spring-managed transactions (through Spring's {@link TransactionSynchronizationManager},
+ * performing the actual cache put/evict/clear operation only in the after-commit phase of a
+ * successful transaction. If no transaction is active, {@link #put}, {@link #evict} and
+ * {@link #clear} operations will be performed immediately, as usual.
+ *
+ * <p>Use of more aggressive operations such as {@link #putIfAbsent} cannot be deferred
+ * to the after-commit phase of a running transaction. Use these with care.
+ *
+ * @author Juergen Hoeller
+ * @author Stephane Nicoll
+ * @author Stas Volsky
+ * @since 5.1
+ * @see TransactionAwareCacheManagerProxy
+ */
+public class TransactionAwareCache implements Cache {
```

Comment:
Looking at the history, I think it would be more logical that `TransactionAwareCacheDecorator` was renamed to `TransactionAwareCache` (still with the 5.1 since given the rename. That would make the git history much more clear.